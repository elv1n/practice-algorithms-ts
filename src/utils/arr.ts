const validateArrIndex = (
  arrLength: number,
  index1: number,
  index2: number
) => {
  if (index1 < 0 || index1 >= arrLength || index2 < 0 || index2 >= arrLength) {
    throw new Error(
      `Indexes should be zero or positive number less than array length. Incoming data: max index ${
        arrLength - 1
      }, index1: ${index1}, index2: ${index2}`
    );
  }
};
export const swap = <T>(arr: T[], fromIndex: number, toIndex: number) => {
  validateArrIndex(arr.length, fromIndex, toIndex);
  [arr[fromIndex], arr[toIndex]] = [arr[toIndex], arr[fromIndex]];
  return arr;
};

export const reorder = <T>(arr: T[], startIndex: number, endIndex: number) => {
  validateArrIndex(arr.length, startIndex, endIndex);
  const [removed] = arr.splice(startIndex, 1);
  arr.splice(endIndex, 0, removed);
  return arr;
};
/**
 * @deprecated
 * new Array(length).fill(value) faster
 */
export const fillArray = <T>(length: number, value: T): T[] => {
  const arr = [];
  for (let i = 0; i < length; i += 1) {
    arr[i] = value;
  }
  return arr;
};
