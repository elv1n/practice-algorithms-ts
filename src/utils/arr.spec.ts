import { fillArray, reorder, swap } from './arr';

describe('arr utils', () => {
  it('should swap numbers', () => {
    expect(swap([1, 2, 3, 4, 5], 2, 0)).toEqual([3, 2, 1, 4, 5]);
  });
  it('should swap strings', () => {
    expect(swap(['1', '2', '3', '4', '5'], 2, 0)).toEqual([
      '3',
      '2',
      '1',
      '4',
      '5',
    ]);
  });
  it('should swap numbers from end to start', () => {
    expect(swap([1, 2, 3, 4, 5], 4, 0)).toEqual([5, 2, 3, 4, 1]);
  });
  it('should swap throw an error for negative number', () => {
    expect(
      swap.bind(null, [1, 2, 3, 4, 5], -1, 0)
    ).toThrowErrorMatchingSnapshot();
    expect(
      swap.bind(null, [1, 2, 3, 4, 5], 0, -1)
    ).toThrowErrorMatchingSnapshot();
  });
  it('should swap throw ean error for larger then max index arguments', () => {
    expect(
      swap.bind(null, [1, 2, 3, 4, 5], 5, 0)
    ).toThrowErrorMatchingSnapshot();
    expect(
      swap.bind(null, [1, 2, 3, 4, 5], 0, 5)
    ).toThrowErrorMatchingSnapshot();
  });
  it('should reorder numbers', () => {
    expect(reorder([1, 2, 3, 4, 5], 4, 0)).toEqual([5, 1, 2, 3, 4]);
  });
  it('should reorder strings', () => {
    expect(reorder(['1', '2', '3', '4', '5'], 2, 0)).toEqual([
      '3',
      '1',
      '2',
      '4',
      '5',
    ]);
  });
  it('should reorder throw an error for negative number', () => {
    expect(
      reorder.bind(null, [1, 2, 3, 4, 5], -1, 0)
    ).toThrowErrorMatchingSnapshot();
    expect(
      reorder.bind(null, [1, 2, 3, 4, 5], 0, -1)
    ).toThrowErrorMatchingSnapshot();
  });
  it('should reorder throw ean error for larger then max index arguments', () => {
    expect(
      reorder.bind(null, [1, 2, 3, 4, 5], 5, 0)
    ).toThrowErrorMatchingSnapshot();
    expect(
      reorder.bind(null, [1, 2, 3, 4, 5], 0, 5)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should fill arr with number', () => {
    expect(fillArray(3, 2)).toEqual([2, 2, 2]);
  });
  it('should fill arr with string', () => {
    expect(fillArray(3, '2')).toEqual(['2', '2', '2']);
  });
});
