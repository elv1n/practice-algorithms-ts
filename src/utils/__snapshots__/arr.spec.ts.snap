// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arr utils should reorder throw an error for negative number 1`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: -1, index2: 0"`;

exports[`arr utils should reorder throw an error for negative number 2`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: 0, index2: -1"`;

exports[`arr utils should reorder throw ean error for larger then max index arguments 1`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: 5, index2: 0"`;

exports[`arr utils should reorder throw ean error for larger then max index arguments 2`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: 0, index2: 5"`;

exports[`arr utils should swap throw an error for negative number 1`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: -1, index2: 0"`;

exports[`arr utils should swap throw an error for negative number 2`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: 0, index2: -1"`;

exports[`arr utils should swap throw ean error for larger then max index arguments 1`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: 5, index2: 0"`;

exports[`arr utils should swap throw ean error for larger then max index arguments 2`] = `"Indexes should be zero or positive number less than array length. Incoming data: max index 4, index1: 0, index2: 5"`;
